File: electron/main/tools/FileTool.ts
import { DynamicStructuredTool } from '@langchain/core/tools';
import { z } from 'zod';
import { promises as fs } from 'fs';
import { join } from 'path';
import { validateReadPath, validateWritePath, validateDirectoryPath } from './pathValidation';

/**
 * FileTool provides safe file system operations to the agent
 */
const MAX_FILE_SIZE = 1024 * 1024; // 1MB limit for file operations
const MAX_WRITE_SIZE = 1024 * 1024; // 1MB limit for write operations
export class FileTool {
  static listDirectory = new DynamicStructuredTool({
    name: 'list_directory',
    description: 'List the contents of a directory. Returns an array of file and directory names.',
    schema: z.object({
      path: z.string().describe('The absolute path of the directory to list'),
    }),
    func: async ({ path }) => {
      // Errors will be caught by the AgentService
      const validatedPath = await validateDirectoryPath(path);
      const items = await fs.readdir(validatedPath, { withFileTypes: true });
      
      const result = await Promise.all(items.map(async item => ({
        name: item.name,
        type: item.isDirectory() ? 'directory' : 'file',
        size: item.isFile() ? (await fs.stat(join(validatedPath, item.name))).size : undefined,
      })));
      
      return JSON.stringify(result, null, 2);
    },
  });

  static readFile = new DynamicStructuredTool({
    name: 'read_file',
    description: 'Read the contents of a text file. Returns the file content as a string.',
    schema: z.object({
      path: z.string().describe('The absolute path of the file to read'),
      encoding: z.enum(['utf-8', 'utf8', 'ascii', 'binary', 'base64', 'hex', 'latin1', 'ucs2', 'utf16le']).optional().default('utf-8').describe('File encoding (default: utf-8)'),
    }),
    func: async ({ path, encoding }) => {
      // Errors will be caught by the AgentService
      const validatedPath = await validateReadPath(path);
      const stats = await fs.stat(validatedPath);
      
      // Check file size limit
      if (stats.size > MAX_FILE_SIZE) {
        throw new Error(`Validation Error: File too large (${Math.round(stats.size / 1024)}KB). Maximum allowed size is ${MAX_FILE_SIZE / 1024}KB.`);
      }
      
      const content = await fs.readFile(validatedPath, encoding as BufferEncoding);
      return content;
    },
  });

  static writeFile = new DynamicStructuredTool({
    name: 'write_file',
    description: 'Write content to a file. Creates the file if it doesn\'t exist, overwrites if it does.',
    schema: z.object({
      path: z.string().describe('The absolute path of the file to write'),
      content: z.string().describe('The content to write to the file'),
      encoding: z.enum(['utf-8', 'utf8', 'ascii', 'binary', 'base64', 'hex', 'latin1', 'ucs2', 'utf16le']).optional().default('utf-8').describe('File encoding (default: utf-8)'),
    }),
    func: async ({ path, content, encoding }) => {
      // Errors will be caught by the AgentService
      const bufferEncoding = Buffer.byteLength(content, encoding as BufferEncoding);
      // Check content size limit
      if (bufferEncoding > MAX_WRITE_SIZE) {
        throw new Error(`Validation Error: Content too large (${Math.round(bufferEncoding / 1024)}KB). Maximum allowed size is ${MAX_WRITE_SIZE / 1024}KB.`);
      }
      
      const validatedPath = await validateWritePath(path);
      await fs.writeFile(validatedPath, content, encoding as BufferEncoding);
      return `Successfully wrote to file: ${validatedPath}`;
    },
  });

  static createDirectory = new DynamicStructuredTool({
    name: 'create_directory',
    description: 'Create a new directory.',
    schema: z.object({
      path: z.string().describe('The absolute path of the directory to create'),
    }),
    func: async ({ path }) => {
      // Errors will be caught by the AgentService
      const validatedPath = await validateWritePath(path);
      await fs.mkdir(validatedPath, { recursive: true });
      return `Successfully created directory: ${validatedPath}`;
    },
  });

  static deleteFile = new DynamicStructuredTool({
    name: 'delete_file',
    description: 'Delete a file from the file system. Use with caution.',
    schema: z.object({
      path: z.string().describe('The absolute path of the file to delete'),
    }),
    func: async ({ path }) => {
      // Errors will be caught by the AgentService
      const validatedPath = await validateReadPath(path);
      await fs.unlink(validatedPath);
      return `Successfully deleted file: ${validatedPath}`;
    },
  });

  /**
   * Get all tools as an array
   */
  static getTools() {
    return [
      this.listDirectory,
      this.readFile,
      this.writeFile,
      this.createDirectory,
      this.deleteFile,
    ];
  }
}

File: backend/tools/file_tool.py
import json
from pathlib import Path
from langchain_core.tools import tool
from tools.path_validation import (
    validate_read_path,
    validate_write_path,
    validate_directory_path,
)

MAX_FILE_SIZE = 1024 * 1024  # 1MB limit
MAX_WRITE_SIZE = 1024 * 1024  # 1MB limit


@tool
def list_directory(path: str) -> str:
    """List the contents of a directory. Returns an array of file and directory names."""
    validated_path = validate_directory_path(path)
    items = []
    
    for item in validated_path.iterdir():
        entry = {
            "name": item.name,
            "type": "directory" if item.is_dir() else "file",
        }
        if item.is_file():
            entry["size"] = item.stat().st_size
        items.append(entry)
    
    return json.dumps(items, indent=2)


@tool
def read_file(path: str, encoding: str = "utf-8") -> str:
    """Read the contents of a text file. Returns the file content as a string."""
    validated_path = validate_read_path(path)
    file_size = validated_path.stat().st_size
    
    if file_size > MAX_FILE_SIZE:
        raise ValueError(
            f"Validation Error: File too large ({file_size // 1024}KB). "
            f"Maximum allowed size is {MAX_FILE_SIZE // 1024}KB."
        )
    
    return validated_path.read_text(encoding=encoding)


@tool
def write_file(path: str, content: str, encoding: str = "utf-8") -> str:
    """Write content to a file. Creates the file if it doesn't exist, overwrites if it does."""
    content_size = len(content.encode(encoding))
    
    if content_size > MAX_WRITE_SIZE:
        raise ValueError(
            f"Validation Error: Content too large ({content_size // 1024}KB). "
            f"Maximum allowed size is {MAX_WRITE_SIZE // 1024}KB."
        )
    
    sanitized_path = validate_write_path(path)
    Path(sanitized_path).write_text(content, encoding=encoding)
    return f"Successfully wrote to file: {sanitized_path}"


@tool
def create_directory(path: str) -> str:
    """Create a new directory."""
    validated_path = validate_write_path(path)
    validated_path.mkdir(parents=True, exist_ok=True)
    return f"Successfully created directory: {validated_path}"


@tool
def delete_file(path: str) -> str:
    """Delete a file from the file system. Use with caution."""
    validated_path = validate_read_path(path)
    validated_path.unlink()
    return f"Successfully deleted file: {validated_path}"


def get_file_tools():
    """Get all file tools as a list."""
    return [list_directory, read_file, write_file, create_directory, delete_file]
